src/Emulator/Extensions/Mocks/HPSHostController.cs

```diff
@@ -171,6 +174,114 @@ private byte[] GetBytesFromSlave(int count, TimeInterval timeInterval)
+        private byte[] GetLongBytes(int count, TimeInterval timeInterval) {            
+            var result = new byte[count];
+            int rounds = count /255;
+            int remainder = count % 255; 
+            for (int i = 0; i <= rounds; i ++) {
+                int toRead = i == rounds ? remainder : 255;
+                Array.Copy(GetBytesFromSlave(toRead, timeInterval), 0, result, i * 255, toRead);
+            }
+            return result;
+        }
+
+        public void StartSocketServer(TimeInterval timeInterval)
+        {  
+            Console.WriteLine(currentSlave.GetType().Name);
+            string path = "/tmp/i2c.sock";
+            if (File.Exists(path))
+            {
+                File.Delete(path);
+            }
+            var socket = new Socket(AddressFamily.Unix, SocketType.Stream, ProtocolType.Unspecified);
+            socket.Bind(new UnixEndPoint(path));
+            socket.Listen(0);
+            Console.WriteLine("Server started waiting for client to connect...");
+            Console.WriteLine("Starting on: "+ path);
+            currentSlave.Read(0);
+            var connection = socket.Accept();
+            Console.WriteLine("Connection established");
+
+            var commandBuffer = new byte[1];
+            int length;
+            while(SocketConnected(connection)){
+                try {
+                    PollForSocketData(connection, (int)numOfBytes.MessageType, commandBuffer);
+                }
+                catch(Exception e)
+                {
+                    Console.WriteLine("{0} Exception caught.", e);
+                    continue;
+                }
+                switch(commandBuffer[0])
+                {
+                    case (byte)SocketI2CMessageType.Stop:
+                        continue;
+                    // Parse read command
+                    case (byte)SocketI2CMessageType.Read:
+                        length = GetLength(connection);
+                    
+                        byte [] dataToSend =  GetLongBytes(length, timeInterval);
+                        try 
+                        {   
+                            connection.Send(dataToSend, 0, length, SocketFlags.None);
+                        }
+                        catch(Exception e)
+                        {
+                            Console.WriteLine("{0} Exception caught.", e);
+                            continue;
+                        }
+                        break;
+                    // Parse write command
+                    case (byte)SocketI2CMessageType.Write:
+                        length = GetLength(connection);
+                        byte[] data = new byte[length];
+                        try 
+                        {   
+                            PollForSocketData(connection, length, data);
+                            if(data.Length == 1 && data[0] == 0x86){
```

Comment:
Ideally this shouldn't be here right? I guess it's just here for experimentation purposes?

Reply:
Ideally no, but seems that there is some latency/lag for the renode side to update their registers unfortunately. @PiotrZierhoffer Do you have any thoughts? This is for polling error register as usually the first read might be 0, but it might not be the case for a second read.